Catch config upload validation errors

craddm opened the pull request:
## :white_check_mark: Checklist

<!--
Thank you for your pull request!

- Before going any further, please review the [guidelines for contributing](../CONTRIBUTING.md) to this repository.
- If you're not yet ready to merge, please mark this pull request as a **draft** and add `'[WIP]'` to the title
- Replace the empty checkboxes [ ] below with checked ones [x] accordingly.
-->

- [x] You have given your pull request a meaningful title (_e.g._ `Enable foobar integration` rather than `515 foobar`).
- [x] You are targeting the appropriate branch. If you're not certain which one this is, it should be **`develop`**.
- [x] Your branch is up-to-date with the **target branch** (it probably was when you started, but it may have changed since then).

### :vertical_traffic_light: Depends on

<!--
List any other PRs that must be merged before this one
-->

### :arrow_heading_up: Summary

<!--
Please explain what your pull request does here.
You might (optionally) want to include screenshots here.
-->

Catches an exception when trying to upload an invalid config file and exits cleanly. Also improves the related error message to make it clearer what the validation error is.

Original error message:
![image](https://github.com/user-attachments/assets/9eabbcdc-cb8e-45dc-8735-0a6c706e0b81)

Catches the error cleanly and yields a clearer error message
![image](https://github.com/user-attachments/assets/ad0a27e8-71ce-4d8e-a1b6-cdfa280c7d63)



### :closed_umbrella: Related issues

<!--
If your pull request will close any open issues (hopefully it will!) then add `Closes #<issue number>` here.
-->

Closes #2210 

### :microscope: Tests

<!--
- Document any manual tests that you have carried out (*e.g.* deploying a new SHM and/or SRE) and confirm which commit you did this with
- Note that automated tests will be run as part of the CI process and will block this PR until they pass.
- Additionally, a successful code review may be required before this PR can be merged.
- If this pull request only changed documentation you can simply state 'Documentation only' here.
-->

Tested locally

craddm committed:
Catch validation errors on config upload

craddm committed:
Provide more readable error message for config validation errors

craddm committed:
Merge branch 'alan-turing-institute:develop' into config-upload-errors

craddm committed:
Fix linting

JimMadge reviewed (commented):
Looks good. I think the output here is clearer than before.

@craddm we could also add a couple of tests here

1. To ensure a config problem give the human-friendly error
2. To check the format of the exception in `yaml_serialisable_model.py`

craddm committed:
Merge remote-tracking branch 'upstream/develop' into config-upload-errors

craddm committed:
Add fixture for config with missing field

craddm committed:
add test for error from missing config field

craddm committed:
make working test of validation error

JimMadge commented on tests/conftest.py:
```diff
@@ -534,6 +534,41 @@ def sre_config_yaml(request):
 
+@fixture
+def sre_config_yaml_missing_field(request):
+    content = """---
+    azure:
+        location: uksouth
+        subscription_id: guid_subscription
+        tenant_id: guid_tenant
+    description: Sandbox Project
+    dockerhub:
+        access_token: dummytoken
+        username: exampleuser
+    name: sandbox
+    sre:
+        admin_ip_addresses:
+        - 1.2.3.4/32
+        data_provider_ip_addresses: []
+        databases: []
+        remote_desktop:
+            allow_copy: false
+            allow_paste: false
+        research_user_ip_addresses: []
+        software_packages: none
+        storage_quota_gb:
+            home: 100
+            shared: 100
+        timezone: Europe/London
+        workspace_skus: []
```
Is there already a fixture with this config (or one that is close enough)? You could use that here instead of duplicating:

```python
@fixture
def sre_config_yaml_missing_field(request, config_yaml):
    config_yaml.replace(
        ...
```

craddm committed:
Add test of log messages from config validation errors

craddm committed:
lint:fmt

craddm committed:
Use replace method to create config with missing field

craddm committed:
Merge remote-tracking branch 'upstream/develop' into config-upload-errors

JimMadge commented:
Looks like a test is trying to use az CLI. Maybe a missing mock fixture.

craddm commented:
> Looks like a test is trying to use az CLI. Maybe a missing mock fixture.

Yes - odd as it was working before we changed things today, and the only thing we changed which touches this test is how the yaml with the missing field is made.

JimMadge commented:
Might be due to a change on develop that was merged in?

craddm commented:
> Might be due to a change on develop that was merged in?

Nope - just tested it out. It works fine if I revert the change in how `sre_config_yaml_missing_field` is made. Can't understand why!

JimMadge commented:
Hmm 🤔 something about how the `sre_config_yaml` fixture is constructed?

JimMadge commented:
Or, the change from missing value to missing field has changed what is happening underneath?

JimMadge commented:
In any case, you should be able to follow the traceback (or code) to find out what is happening.

craddm committed:
Correct removal of text from example SRE config

craddm commented:
We just weren't doing it correctly, so the `yaml` wasn't actually invalid and the `upload` function was progressing to the point where it needed to be logged in. The validation errors happen before that point, so it isn't necessary to do extra mocks to cope with logins etc.

JimMadge commented:
Oh, that is brilliant. We were too good at _not_ writing bad YAML 😆.
